src/lem/constrainer.rs

```diff
@@ -90,28 +148,226 @@ impl LEM {
         }
```

Comment:
Lack of a concrete path means you are on a concrete path?! I find this abstraction confusing at best. I suspect it would be better to convert this `Option<Boolean>` into a three-valued enum with very clear names and some documentation explaining exact significance.

Reply:
GitHub is not doing a good job of displaying this comment and its context, but these comments relate the definition of `on_concrete_path()`.

Reply:
Sorry, I was confused by bool vs Boolean — but I continue to find this sketchy. In particular, I suspect the enduring confusion around bellperson use of Option vs application-level optionality may be biting you here. Will keep trying to parse this.

Reply:
Remember, if bellperson's `get_value()` returns `None`, that means you are synthesizing a circuit rather than generating a witness. My reading of this code is therefore that it's meaningless to ask whether you're on a concrete path during circuit synthesis. Confusingly, a `concrete_path` of `None` means this *is* a concrete path. If all that is what's actually intended, okay… I'll try to square that with usage as I continue. If not, please clarify (which might require code changes).

Reply:
Okay, I've now encountered the documentation of this, which I realize was not part of this PR — but nevertheless I need to come up to speed on. That clarifies, and addresses my questions/concerns above.

Reply:
I think in this case, for all the reasons I was confused when trying to read this code, though, I would strongly consider something like an enum with variants like `AlwaysConcrete` and `MaybeVirtual` or something like that.